### ISSUE TITLE
Egress gateway doesn't seem to be working in 1.15, after upgrade from 1.13, 1.14

### ISSUE THREAD

**Issue by nward** [cilium/cilium#32971]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I have been using 1.13 until recently, using a egress gateway policy that specifies an address that is not locally configured in the kernel.

This works for outbound, I attract return traffic to the node with a local pod backing a load balancer service with the SNAT IP specified, so that it advertises the IP through BGP.

This works great, though is a bit of a hack to attract the traffic I admit!
I get a message in the logs saying it's unable to find the IPv4 address on any interfaces, but it works fine all the same.

My use case is a single pod which needs to use this specific outbound IP address. I have several pods like this which each require different egress IPs.

Since upgrading to 1.15, the egress gateway policy doesn't appear to work - either with this non-connected SNAT IP hack, OR when specifying the outgoing IP of the node, or the interface name.

I have noted that I have the above unusual configuration, because ideally that's what I'm trying to get working again. Reading the code (I'm not a go coder, so may be missing something) it looks like it needs the address attached on the outgoing interface on the node, which is not ideal for my use case. So far as I can tell, this was always the case though.

In my dev environment where I have reproduced this, I have host routing set to legacy, however in other environments I have set this to BPF with no change in behaviour.

### Cilium Version

Observed on both of:

1.15.1 (rke2 1.26.14+rke2r1)
1.15.5 (rke2 v1.28.10+rke2r1)

### Kernel Version

Observed on both of:

Linux hostname 5.17.0-1020-oem #21-Ubuntu SMP PREEMPT Fri Oct 14 09:33:24 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

Linux hostname 6.8.0-35-generic #35-Ubuntu SMP PREEMPT_DYNAMIC Mon May 20 15:51:52 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Both of
v1.26.14+rke2r1
v1.28.10+rke2r1

### Regression

This worked in 1.14.4, I tested it on rke2 1.26.13+rke2r1, which was the release before cilium 1.15.1 was bundled.

### Sysdump

[cilium-sysdump-20240608-205019.zip](https://github.com/user-attachments/files/15747347/cilium-sysdump-20240608-205019.zip)

### Relevant log output

```shell
❯ kubectl -n kube-system exec ds/cilium -- cilium-dbg bpf egress list
Defaulted container "cilium-agent" out of: cilium-agent, install-portmap-cni-plugin (init), config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
No entries found.
```

### Anything else?

The policy I have configured is as below:
```
- apiVersion: cilium.io/v2
  kind: CiliumEgressGatewayPolicy
  metadata:
    creationTimestamp: "2023-07-21T10:09:02Z"
    generation: 8
    labels:
      kustomize.toolkit.fluxcd.io/name: apps
      kustomize.toolkit.fluxcd.io/namespace: flux-system
    name: mta-egress-0
    resourceVersion: "315439081"
    uid: aa83691a-1d28-4d9b-a1ee-ec6bfe0799dd
  spec:
    destinationCIDRs:
    - 0.0.0.0/0
    egressGateway:
      egressIP: 103.247.72.89
      nodeSelector:
        matchLabels:
          kubernetes.io/hostname: email-worker-1
    selectors:
    - podSelector:
        matchLabels:
          statefulset.kubernetes.io/pod-name: mta-0
```

With the address noted configured locally I do not get an error message in the agent logs, but, the gateway does not appear to be set up either.

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by nward** [comment]
I forgot to note - my ideal use case may be related to #30747, however I am unable to get egress gateway policy to work at all, so raised this as a unique issue.

**Comment by julianwiedmann** [comment]
Yep it's possible that such "exotic" setups no longer play well with the latest EGW stabilization efforts. [The docs](https://docs.cilium.io/en/v1.15/network/egress-gateway/) list two requirements in particular:

> The egress IP must be assigned to a network device on the node.

and 

>For gateway nodes with multiple network interfaces, Cilium selects the egress network interface based on the node’s routing setup (ip route get <externalIP> from <egressIP>).

**Comment by nward** [comment]
Exotic is a good word for it :-)
It was a nice solution though, because I don't have to configure the IP address on the node, and I can move the pod, service, and egress gateway policy around to different nodes as I need. It would be great to have that capability back in future versions - even better if it can be done without having to have a dummy pod/service :-)


Besides the exotic approach I was taking - on 1.15 I am unable to get egress policy to work at all. I have tried all of:
1) set the /32 on a dummy interface (which passes the `ip route get from` test)
2) set the /32 on the only ethernet interface (which passes the `ip route get from` test)
3) modify the egress policy to use an egress IP of the only ethernet interface instead of a /32
4) modify the egress policy to use the interface name of the only ethernet interface instead of the IP

The egress gateway policy I'm using above is identical to the one that works for me in 1.14.x and below, apart from the IP/interface in tests 3 and 4. The node and pod selectors are identical.

In all of the above cases, the egress gateway policy is not installed at all, and the packets leave the node with the matched pod as though no egress gateway policy is installed at all. The node that is set as an egress gateway for the pod does not forward any of those packets.

In short - I am unable to get egress gateway policy to work at all, even in "normal" configurations.

**Comment by julianwiedmann** [comment]
I'd suggest starting from https://docs.cilium.io/en/v1.15/network/egress-gateway/#troubleshooting, to understand how the policy gets plumbed into the datapath.

**Comment by nward** [comment]
Thanks - have already been through the troubleshooting when comparing the various versions, I simply get `No entries found.` on 1.15, and am unable to find any references to the policy in any state data I can find.

Have tried running bpf egress list on each node, rather than just whichever kubectl selects, and get the same results.

**Comment by julianwiedmann** [comment]
>I simply get No entries found. on 1.15, and am unable to find any references to the policy in any state data I can find.

If `bpf egress list` gives you no entries, your `podSelector` isn't set up correctly.

**Comment by nward** [comment]
Thanks @julianwiedmann. The podSelector is not changing between what is applied to 1.14.x and 1.15.x. The pod selector matches a pod, and the node selector matches a node, both correctly and as expected.

The below is in my dev cluster which is on cilium 1.15.5. My test cluster which is on 1.14.x has egress gateway policies which are largely the same.

This specific egress gateway policy should be telling the pod mta-1 to egress to the public Internet through the only interface IP on email-worker-2. The pod matching the selector is on the node email-worker-1, while the selected node with the egressIP is email-worker-2.

If you have further suggestions for how to debug this I would be all ears and very grateful!

```
Name:         mta-egress-1
Namespace:
Labels:       kustomize.toolkit.fluxcd.io/name=apps
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  cilium.io/v2
Kind:         CiliumEgressGatewayPolicy
Metadata:
  Creation Timestamp:  2023-07-21T10:09:02Z
  Generation:          3
  Resource Version:    316082523
  UID:                 592e38c8-ac84-49ae-b1aa-40808348b1cd
Spec:
  Destination CID Rs:
    0.0.0.0/0
  Egress Gateway:
    Egress IP:  10.1.2.3
    Node Selector:
      Match Labels:
        kubernetes.io/hostname:  email-worker-2
  Selectors:
    Pod Selector:
      Match Labels:
        statefulset.kubernetes.io/pod-name:  mta-1
```

```
❯ kubectl get nodes -l kubernetes.io/hostname=email-worker-2
NAME             STATUS   ROLES    AGE    VERSION
email-worker-2   Ready    <none>   527d   v1.28.10+rke2r1

❯ kubectl get pods -o wide -l statefulset.kubernetes.io/pod-name=mta-1
NAME    READY   STATUS    RESTARTS   AGE   IP            NODE             NOMINATED NODE   READINESS GATES
mta-1   2/2     Running   0          8d    10.42.3.155   email-worker-1   <none>           <none>

❯ kubectl -n kube-system exec ds/cilium -- cilium-dbg bpf egress list
Defaulted container "cilium-agent" out of: cilium-agent, install-portmap-cni-plugin (init), config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
No entries found.
```

```
ubuntu@email-worker-2:~$ ip addr show dev ens3
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:54:00:62:8b:97 brd ff:ff:ff:ff:ff:ff
    altname enp0s3
    inet 10.1.2.3/24 metric 100 brd 10.1.2.255 scope global dynamic ens3
       valid_lft 2337sec preferred_lft 2337sec
    inet6 fe80::5054:ff:fe62:8b97/64 scope link
       valid_lft forever preferred_lft forever
ubuntu@email-worker-2:~$ ip route get 4.1.1.1 from 10.1.2.3
4.1.1.1 from 10.1.2.3 via 10.1.2.1 dev ens3 uid 1000
    cache
```

**Comment by julianwiedmann** [comment]
Ah, you're using `statefulset.kubernetes.io/pod-name` 💡.

Then most likely you got broken by what's described in https://github.com/cilium/cilium/pull/28003#issuecomment-2084050409. EGW only considers identity-relevant labels.

**Comment by nward** [comment]
Hi @julianwiedmann - ahhh yes that's exactly what is happening. Great spotting.

I've configured Kyverno to add an additional label to the pods in question now and I'm using that in the egress gateway policy - and all works exactly as I want.

Even works well with my somewhat alternative way of using the IPs without having them configured on actual interfaces.

I will mark this as closed, thank you for your help and patience on this.
